### PR TITLE
Added quality of service(QOS) via Differentiated services

### PR DESCRIPTION
**Pull request by nikkytub**

### CODE REVIEW

**Comment by jafingerhut** [comment]
Wow.  It is uncommon for a PR adding a new worked-out exercise here.  Interesting.

@jnfoster Do you know who should review such things?  Perhaps this is an invited submission from someone planning to give P4 tutorials some time in the near future?

**Comment by sanooprajan** [comment]
is it not better to do the classification in the ingress itself if all you do is just a classifcation ?

**Comment by nikkytub** [comment]
> is it not better to do the classification in the ingress itself if all you do is just a classifcation ?

Done:)

**Comment by nikkytub** [comment]
> Wow. It is uncommon for a PR adding a new worked-out exercise here. Interesting.
> 
> @jnfoster Do you know who should review such things? Perhaps this is an invited submission from someone planning to give P4 tutorials some time in the near future?

Thanks @jafingerhut I like the language and the idea of data plane programming and would like to actively participate in it.

**Comment by jnfoster** [comment]
@nikkytub This looks great! Assuming you were not previously a member of the P4 Language Consortium, please send an email to me and we'll get things set up for contributing to this repository...

**Comment by nikkytub** [comment]
> @nikkytub This looks great! Assuming you were not previously a member of the P4 Language Consortium, please send an email to me and we'll get things set up for contributing to this repository...

@jnfoster Thanks. Done:)

File: exercises/qos/README.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+program.
+
+> **Spoiler alert:** There is a reference solution in the `solution`
+> sub-directory. Feel free to compare your implementation to the reference.
+
+## Step 1: Run the (incomplete) starter code
+
+The directory with this README also contains a skeleton P4 program,
+`qos.p4`, which initially implements L3 forwarding. Your job (in the
+next step) will be to extend it to properly append set the `diffserv` bits.
```
Comment (jafingerhut) [inline exercises/qos/README.md]: Not sure of the intended meaning of this sentence.  Perhaps the intent is "will be to extend it to properly set the `diffuser` bits.", removing the word "append"?

File: exercises/qos/README.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+2. Parsers for Ethernet, IPv4,
+3. An action to drop a packet, using `mark_to_drop()`.
+4. An action (called `ipv4_forward`), which will:
+	1. Set the egress port for the next hop.
+	2. Update the ethernet destination address with the address of
+           the next hop.
+	3. Update the ethernet source address with the address of the switch.
+	4. Decrement the TTL.
+5. An ingress control block that checks the protocols and sets the ipv4.diffserv.
+6. A deparser that selects the order in which fields inserted into the outgoing
```
Comment (jafingerhut) [inline exercises/qos/README.md]: Maybe replace "in which fields inserted" with "in which headers are inserted" ?

File: exercises/qos/README.md (hunk 1 above)
Comment (nikkytub) [reply exercises/qos/README.md]: Done

File: exercises/qos/README.md (hunk 2 above)
Comment (nikkytub) [reply exercises/qos/README.md]: Done

**Comment by nikkytub** [comment]
Hi @jafingerhut @jnfoster, can you please merge it?

**Comment by jnfoster** [comment]
Thanks for the contribution!

**Comment by nikkytub** [comment]
> Thanks for the contribution!

Thanks a ton. You are most welcome:)

**Pull request merged**